objects/obj_pnunit/Alarm_5.gml

```diff
@@ -1,95 +1,97 @@
-                    if (cur_exp<20) then new_exp+=10;
-                }
-                if (new_exp>0){
-                    unit.add_exp(new_exp);
-                    obj_ncombat.total_battle_exp_gain+=new_exp;
-                }
-                if (unit.IsSpecialist("libs")) then unit.update_powers();
-                // Need some kind of report here
-            }
+if (obj_ncombat.defeat == 0) {
+    obj_ncombat.total_battle_exp_gain = obj_ncombat.threat * 50;
+    var _current_exp;
+    var _eligible_units = [];
+    var _exp_mod = 1;
+    var _unit_xp_data = [];
+    var _unit_recovery_score = obj_ncombat.unit_recovery_score;
+
+    for (var i = 0; i < array_length(unit_struct); i++) {
+        _unit = unit_struct[i];
+        if (is_struct(_unit) && ally[i] == false) {
+            if (marine_dead[i] == 1 && marine_type[i] != "" && _unit_recovery_score > 0) { // Apothecaries saving marines
+                obj_ncombat.unit_recovery_score -= 1;
+                _unit.update_health(irandom(20) - 10);
+                marine_dead[i] = 0;
+                obj_ncombat.units_saved += 1;
+            } 
             
-            if (marine_type[i]!="") and (marine_dead[i]==1) and (ally[i]=false){
-                if (marine_type[i]=="Chapter Master"){
-                    if (obj_ncombat.unit_recovery_score>0){
-                        obj_ncombat.unit_recovery_score-=1;
-                        unit.update_health(irandom(20)-10);
-                        marine_dead[i]=0;
-                        obj_ncombat.units_saved+=1;                
-                    }
-                }else if (marine_type[i]!=""){
-                    if (obj_ncombat.unit_recovery_score>0){
-                        obj_ncombat.unit_recovery_score-=1;
-                        unit.update_health(irandom(20)-10);
-                        marine_dead[i]=0;
-                        obj_ncombat.units_saved+=1;
-                        // show_message(string(marine_type[i])+" is saved by an apothecary");
-                    }
-                }
+            if (marine_dead[i] == 0) { // EXP allocation
+                _current_exp = _unit.experience;
+
+                _exp_mod = 1 - (_current_exp / 200);
+
+                _unit_xp_data = [_unit, _exp_mod];
+                array_push(_eligible_units, _unit_xp_data);
+            }
+        }
+    }
+
+    // EXP allocation
+    var _eligible_units_count = array_length(_eligible_units);
+    if (_eligible_units_count > 0 && obj_ncombat.total_battle_exp_gain > 0) {
+        var _individual_exp = round(obj_ncombat.total_battle_exp_gain / _eligible_units_count);
+        obj_ncombat.average_battle_exp_gain = _individual_exp;
+        for (var i = 0; i < _eligible_units_count; i++) {
+            var _unit = _eligible_units[i][0];
+            var _exp_mod = _eligible_units[i][1];
+            var _exp_update_data = _unit.add_exp(round(_individual_exp*_exp_mod));
```

Comment:
> we support fragment exp as it's required for multi turn exp growth, no reason to round here therefore as it's inconsistent imo

I remember I did that, because after a battle I saw decimal exp values, which I never saw before this PR. Hence, assumed that all existing exp additions are rounded before adding, while the display is not.
Not sure what exactly is wrong, me misremembering how EXP looks or displayed EXP not being rounded.
I'll check this.